### ISSUE TITLE
Extend no-multi-spaces with allow alignment option

### ISSUE THREAD

**Issue by neverfox** [eslint/eslint#1524]
Is it possible to have a rule to allow alignment of `=` (which introduces multi spaces) without having to turn off `no-multi-spaces`?

The following would be considered a warning:

```javascript
var a   = 1;

var a    = 1;
var bar = 1;
```

The following would not:
```javascript
var a   = 1;
var bar = 1;
```

Currently `no-multi-space` complains about what ESLint calls good coding conventions:

```javascript
// Good
var count   = 10,
    name    = "Nicholas",
    found   = false,
    empty;
```

I agree that this is good, but turning off `no-multi-spaces` completely is a blunt instrument that would hide warnings you want.

**Comment by dwelle** [comment]
:+1: I have the same issue with this rule, while disabling it around each and every declaration block doesn't sound like a good idea, as @nzakas suggests.

**Comment by dmmalam** [comment]
+1

**Comment by QuentinRoy** [comment]
+1

**Comment by ivanoats** [comment]
I am also wondering what to do here, and agree with @dwelle that I don't want to disable around every block.

**Comment by nzakas** [comment]
We might be able to do this with the change @btmills made to the rule.

**Comment by btmills** [comment]
@nzakas good point! This should permit vertical alignment in variable declarations:


```js
"key-spacing": [2, { "exceptions": {
    "VariableDeclarator": true
}]
```

**Comment by gilbarbara** [comment]
Hey @nzakas @btmills 

I just saw that `VariableDeclarator` exception is used in `no-multi-spaces` test file but it's not implemented in the rule itself.
Are you going to implement? Are there any other rules that has multiple exceptions for me to use as an example and try to do it myself?

Thanks

**Comment by nzakas** [comment]
I'm not sure what you're asking. What @btmills described is already implemented. It's here: https://github.com/eslint/eslint/blob/master/lib/rules/no-multi-spaces.js#L22

**Comment by gilbarbara** [comment]
Is this the proper use?
```javascript
"no-multi-spaces": [1, {
    "exceptions": {
        "VariableDeclaration": true
    }
}]
```
It's not working for me, I still get the "Multiple spaces found before '='"

**Comment by btmills** [comment]
@gilbarbara make sure you're using `VariableDeclarator` and not `VariableDeclaration`. The initialization is part of the declarator.

**Comment by gilbarbara** [comment]
@btmills YES! :+1: 
Thanks a lot!

**Comment by nzakas** [comment]
Can we close this?

**Comment by gilbarbara** [comment]
Yes, definitely!

**Comment by martinstreicher** [comment]
Is there a way to allow the alignment of `=`s in default values for args? 

```
function (
  x   = 1,
  abc = 2) { 
```
?

**Comment by btmills** [comment]
If you plug that source code into the [parser demo](http://eslint.org/parser/), you should see that `AssignmentPattern` is the exception you want.

**Comment by martinstreicher** [comment]
Yep, I just found that reference and was able to set the option. Thanks!

**Comment by akre54** [comment]
What about for aligning literals in function calls?

```js
foo('one',   'two',   'three')
foo('four',  'five',  'six')
foo('seven', 'eight', 'nine')
```

I think there has to be a smarter option than just ignoring a whole class of node for this type of alignment. Shouldn't it be possible to compare the `start` prop of the AST?

**Comment by alexilyaev** [comment]
@btmills I can't find the right exception for this code (using Class properties through babel preset `stage-1`):

```
class foo {
  one  = 1;
  four = 4;
}
```

The online parser can't read that obviously.
Is there an exception keyword I can use to tell `no-multi-spaces` to ignore that?

Thanks.

**Comment by btmills** [comment]
Try [astexplorer.net](http://astexplorer.net/) to inspect the syntax tree from a number of popular parsers. @alexilyaev in your case with babel-eslint, it looks like you want `ClassProperty`.

**Comment by alexilyaev** [comment]
@btmills Alright, that actually worked!
Thanks.
